src/main/scala/is/hail/io/vcf/ExportVCF.scala

```diff
@@ -1,96 +1,106 @@
-    case TArray(elementType, _) => "."
+    case TArray(_, _) => "."
+    case TSet(_, _) => "."
     case _ => "1"
   }
 
-  def strVCF(sb: StringBuilder, elementType: Type, a: Annotation) {
-    if (a == null)
-      sb += '.'
-    else {
-      elementType match {
-        case TFloat32(_) =>
-          val x = a.asInstanceOf[Float]
-          if (x.isNaN)
-            sb += '.'
-          else
-            sb.append(x.formatted("%.5e"))
-        case TFloat64(_) =>
-          val x = a.asInstanceOf[Double]
-          if (x.isNaN)
-            sb += '.'
-          else
-            sb.append(x.formatted("%.5e"))
-        case TInt64(_) =>
-          val x = a.asInstanceOf[Long]
-          if (x > Int.MaxValue || x < Int.MinValue)
-            fatal(s"Cannot convert Long to Int if value is greater than Int.MaxValue (2^31 - 1) or less than Int.MinValue (-2^31). Found $x.")
-          sb.append(elementType.str(x))
-        case _ => sb.append(elementType.str(a))
-      }
+  def strVCF(sb: StringBuilder, elementType: Type, m: MemoryBuffer, offset: Long) {
+    elementType match {
+      case TInt32(_) =>
+        val x = m.loadInt(offset)
+        sb.append(x.toString)
+      case TInt64(_) =>
+        val x = m.loadLong(offset)
+        if (x > Int.MaxValue || x < Int.MinValue)
+          fatal(s"Cannot convert Long to Int if value is greater than Int.MaxValue (2^31 - 1) " +
+            s"or less than Int.MinValue (-2^31). Found $x.")
+        sb.append(x.toInt.toString)
+      case TFloat32(_) =>
+        val x = m.loadFloat(offset)
+        if (x.isNaN)
+          sb += '.'
+        else
+          sb.append(x.formatted("%.5e"))
+      case TFloat64(_) =>
+        val x = m.loadDouble(offset)
+        if (x.isNaN)
+          sb += '.'
+        else
+          sb.append(x.formatted("%.5e"))
+      case TString(_) =>
+        sb.append(TString.loadString(m, offset))
+      case TCall(_) =>
+        val p = Genotype.gtPair(m.loadInt(offset))
+        sb.append(s"${ p.j }/${ p.k }")
+      case _ =>
+        fatal(s"VCF does not support type $elementType")
     }
   }
-
-  def emitFormatField(f: Field, sb: StringBuilder, a: Annotation) {
-    f.typ match {
-      case TCall(_) => sb.append(Call.toString(a.asInstanceOf[Call]))
-      case it: TIterable =>
-        if (a == null)
+  
+  def iterableVCF(sb: StringBuilder, t: TIterable, m: MemoryBuffer, length: Int, offset: Long) {
+    if (length > 0) {
+      var i = 0
+      while (i < length - 1) {
+        if (t.isElementDefined(m, offset, i)) {
+          val eOffset = t.loadElement(m, offset, length, i)
+          strVCF(sb, t.elementType, m, eOffset)
+        } else
           sb += '.'
-        else {
-          val arr = a.asInstanceOf[Iterable[_]]
-          arr.foreachBetween(a => strVCF(sb, it.elementType, a))(sb += ',')
-        }
-      case t => strVCF(sb, t, a)
-    }
+        sb += ','
+        i += 1
+      }
+      if (t.isElementDefined(m, offset, i)) {
```

Comment:
This is unnecessary duplication.  Just do something like `if (i > 0) sb += ','` at the top of the loop (or the analogous condition at the bottom).